File: go/lc503.go
package main

func nextGreaterElements(nums []int) []int {
	n := len(nums)
	stack, ans := make([]int, n), make([]int, n)
	for i := range ans {
		ans[i] = -1
	}

	for cnt := 0; cnt < 2; cnt++ {
		for i := 0; i < n; i++ {
			for len(stack) > 0 && nums[i] > nums[stack[len(stack)-1]] {
				ans[stack[len(stack)-1]] = nums[i]
				stack = stack[:len(stack)-1]
			}
			stack = append(stack, i)
		}
	}

	return ans
}

File: java/com/lc/lc503.java
package lc;

import java.util.*;

public class lc503 {
    public int[] nextGreaterElements(int[] nums) {
        Deque<Integer> stack = new LinkedList<>();
        int[] ans = new int[nums.length];

        for (int i = 0; i < nums.length; i++){
            ans[i] = -1;
        }

        for (int cnt = 0; cnt < 2; cnt++){
            for (int i = 0; i < nums.length; i++){
                while (!stack.isEmpty() && nums[i] > nums[stack.peek()]){
                    ans[stack.poll()] = nums[i];
                }
                stack.push(i);
            }
        }

        return ans;
    }
}